From a6c69b98e3d75324e13a5fe9e90013ce91f35d9b Mon Sep 17 00:00:00 2001
From: Zhen Lei <thunder.leizhen@huawei.com>
Date: Wed, 27 Apr 2022 18:48:37 +0800
Subject: [PATCH] of: Enable crashkernel=X,high to support kexec option -s
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: a6c69b98e3d75324e13a5fe9e90013ce91f35d9b
Modified-by-SEL: Yes, modified due to different context


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I545H8
CVE: NA

-------------------------------------------------------------------------

For "crashkernel=X,high", there must be two crash regions: high=crashk_res
and low=crashk_low_res. But now the syscall kexec_file_load() only add the
crashk_res into "linux,usable-memory-range", this causes the second kernel
to have no available dma memory. Fix it like kexec tool do for option -c.

Signed-off-by: Zhen Lei <thunder.leizhen@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/of/kexec.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/of/kexec.c b/drivers/of/kexec.c
index f335d941a716..d8231c34e873 100644
--- a/drivers/of/kexec.c
+++ b/drivers/of/kexec.c
@@ -396,6 +396,15 @@ void *of_kexec_alloc_and_setup_fdt(const struct kimage *image,
 				crashk_res.end - crashk_res.start + 1);
 		if (ret)
 			goto out;
+
+		if (crashk_low_res.end) {
+			ret = fdt_appendprop_addrrange(fdt, 0, chosen_node,
+					FDT_PROP_MEM_RANGE,
+					crashk_low_res.start,
+					crashk_low_res.end - crashk_low_res.start + 1);
+			if (ret)
+				goto out;
+		}
 	}
 
 	/* add bootargs */
-- 
2.34.1

